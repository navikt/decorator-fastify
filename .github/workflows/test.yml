name: test

on:
  pull_request:
    branches: 
      - main

env:
  IMAGE: ghcr.io/${{ github.repository }}:${{ github.sha }}
jobs:
  test:
    name: test
    runs-on: ubuntu-latest
    permissions:
      contents: read
      id-token: write
      packages: write
    steps:
      - uses: actions/checkout@v3
      - uses: oven-sh/setup-bun@v1
        with:
          bun-version: '1.0.1'
      - name: Install dependencies
        run: bun install --immutable
        env:
          NODE_AUTH_TOKEN: ${{ secrets.READER_TOKEN }}
      - name: Build app
        run: bun run build
      - name: Run tests
        run: bun test