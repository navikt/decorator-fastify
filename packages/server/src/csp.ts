import {
    CSPDirectives,
    UNSAFE_EVAL,
    UNSAFE_INLINE,
    BLOB,
    DATA,
    getCSP,
} from "csp-header";
import { Handler, responseBuilder } from "./lib/handler";
import { env } from "./env/server";

const navNo = "*.nav.no";
const cdnNavNo = "cdn.nav.no";

const vergicScreenSharing = "*.psplugin.com";
const vergicDotCom = "www.vergic.com"; // seems to only be used for a single placeholder image
const boostChatbot = "*.boost.ai";
const boostScript =
    process.env.ENV === "prod" ? "nav.boost.ai" : "navtest.boost.ai";
const vimeoPlayer = "player.vimeo.com"; // used for inline videos in the chat client
const qbrick = "video.qbrick.com"; // used for inline videos in the chat client
const vimeoCdn = "*.vimeocdn.com"; // used for video preview images

const hotjarCom = "*.hotjar.com";
const hotjarIo = "*.hotjar.io";
const taskAnalytics = "*.taskanalytics.com";
const googleFonts = "*.googleapis.com";
const googleFontsStatic = "*.gstatic.com";

const styleSrc = [
    navNo,
    vergicScreenSharing,
    UNSAFE_INLINE, // various components with style-attributes,
    googleFonts,
    googleFontsStatic,
];

const scriptSrc = [
    navNo,
    vergicScreenSharing,
    hotjarCom,
    taskAnalytics,
    boostScript,
    // localhost testing
    UNSAFE_INLINE, // vergic, hotjar
];

const workerSrc = [
    navNo,
    BLOB, // vergic
];

// @TODO: Merge with bun responses

const directives: Partial<CSPDirectives> = {
    "default-src": [navNo],
    "script-src": [
        ...scriptSrc,
        UNSAFE_EVAL, // vergic
    ],
    "script-src-elem": scriptSrc,
    "worker-src": workerSrc,
    "child-src": workerSrc, // for browsers lacking support for worker-src
    "style-src": styleSrc,
    "style-src-elem": styleSrc,
    "font-src": [
        vergicScreenSharing,
        hotjarCom,
        cdnNavNo,
        googleFonts,
        googleFontsStatic,
        DATA, // ds-css
    ],
    "img-src": [navNo, vergicScreenSharing, vimeoCdn, hotjarCom, vergicDotCom],
    "frame-src": [hotjarCom, vimeoPlayer, qbrick, navNo],
    "connect-src": [
        navNo,
        boostChatbot,
        vergicScreenSharing,
        hotjarCom,
        hotjarIo,
        taskAnalytics,
    ],
};

const localDirectives = Object.entries(directives).reduce(
    (acc, [key, value]) => {
        return {
            ...acc,
            [key]: Array.isArray(value) ? [...value, "localhost:* ws:"] : value,
        };
    },
    {},
);

export const cspDirectives =
    env.ENV === "localhost" ? localDirectives : directives;
export const csp = getCSP({
    presets: [cspDirectives],
});

export const cspHandler: Handler = {
    method: "GET",
    path: "/api/csp",
    handler: () => {
        return responseBuilder().json(cspDirectives).build();
    },
};
