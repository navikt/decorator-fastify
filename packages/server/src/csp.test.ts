import { expect, test } from 'bun:test';
import { cspHandler } from './csp';

test('CSP headers match snapshot', async () => {
    // Dummy request
    const cspHeaders = cspHandler.handler({
        request: new Request('https://localhost:3000/api/csp'),
        url: new URL('https://localhost:3000/api/csp'),
        query: {}
    });

    expect(cspHeaders).toMatchSnapshot();
});
