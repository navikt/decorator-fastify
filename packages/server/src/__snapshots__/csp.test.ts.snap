// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`CSP headers match snapshot 1`] = `Promise {}`;
